manpages/mrcx.man
.na
.nh
.TH MRCX 1 2.40 BL3DEMC
.SH NAME
mrcx \- Convert MRC files between different operating systems

.SH SYNOPSIS
mrcx [-vms|-ieee] filename [optional_output_filename]
.SH DESCRIPTION
Numbers are stored on computers in two different orders, termed
"little-endian" (as on Windows and Linux PCs and Macs with Intel processors)
and "big-endian" (as on Macs with PowerPC processors and some older machines).
Moreover, Unix machines store floating-point numbers in IEEE
format, while VMS machines stored them in a slightly different format.  Mrcx
can convert MRC image files from one machine's format to another by swapping
bytes and, when necessary, converting the format of the floating-point
numbers.
.PP
The type of data conversion is specified with the
.B -vms
or
.B -ieee
option.  (The "Usage" statement issued by entering mrcx without arguments will
indicate which is the default, probably ieee).  Namely,
.B -vms
will convert files between VMS format and big-endian Unix format, while
.B -ieee
will convert files between little-endian and big-endian Unix formats.
.PP
If an output filename is given, then the converted data are written to a new
file; otherwise they are written back into the input file.  Conversion is
completely reversible, so if an erroneous command is given to convert a file
in place, just issue the same command to undo it.

.SH AUTHORS
Ross Dargahi, Jim Kremer, and David Mastronarde

.SH BUGS
Extra data in the extended image header are converted as short integers, which
works for data from SerialEM but not for extended headers in the FEI/Agard
format.

Conversion to and from VMS format is available only on big-endian machines.

There is no longer an attempt to write a new title for the conversion.

Email bug reports to mast@colorado.edu.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/6/13: Added option to Imodinfo to get contour lengths broken out by
fine-grained color, and made -o option provide object list for all outputs.

12/5/13: Updated the eTomo manpage.  Improved the eTomo help statement.
(Bug# 1764)

12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)


mrc/mrcx.c
/*
 *  mrcx.c -- Convert between Vax, SGI, and PC data formats
 *
 *  Original author: Ross Darghi, Integrated Crystollagraphy Environment
 *  Revised by: Jim Kremer, David Mastronarde   email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2005 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id$
 */

/* 
 *
 * PURPOSE....:
 *   To convert an MRC file in VAX format to/from big endian/IEEE format, or
 *   a file in big endian IEEE to/from little endian IEEE format. Byte,
 *   float (4 byte), short int (2 byte), complex float (8 byte), RGB (3 byte),
 *   and complex short (4 byte) types are supported.
 *
 * REVISIONS..:
 *   Ross Dargahi     April 1990
 *
 *   Jim Kremer       Modified July 1993 for HVEM 3D mrc files.
 *                    Modified September 1993 for loading in HVEM extra data.
 *
 *                    Rewrote and renamed program to mrcx.  Program now uses
 *                    the mrcfiles library. Feb 1994
 *                    Jan   1995.  Byte data converted inplace is skipped.
 *                                 Large images don't crash program.
 *                    April 1995   Added bit mode 9-15.
 *                                 Fixed bug that didn't copy byte data.
 *                    Nov   1995   Fixed bug that destroyed header if data 
 *                                 type was unknown.
 *
 *  David Mastronarde Oct   2000   Rewrote to handle either VMS/Unix or
 *                                 big/little IEEE conversions, and to detect
 *                                 which direction to do the conversion in.
 *                                 Eliminated bit modes as hopeless.
 ******************************************************************************/

/******************************************************************************
 *   Include Files
 ******************************************************************************/
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
/*#include <fcntl.h> */
#include <math.h>

/*#include <environ.h> */
#include "mrcfiles.h"
#include "b3dutil.h"

#if defined(WIN32_BIGFILE) || defined(MAC103_BIGFILE)
#define fseek b3dFseek 
#define fread b3dFread 
#define fwrite b3dFwrite 
#define rewind b3dRewind
#endif

/* prototypes */
void convertBytes(FILE  *fp, b3dUByte *buffer, int   noBytes, int direction); 
void convertShorts(FILE  *fp, b3dUByte *buffer, int   noShorts, int direction);
void convertFloats(FILE  *fp, b3dUByte *buffer, int   noFloats, int direction);
void convertLongs(FILE  *fp, b3dUByte *buffer, int   noLongs, int direction); 
void convertBody(void (*convertFunc)(), struct MRCheader *header, int size,
                 int factor, int direction, FILE *infp, FILE *outfp);
void convertHeader(void (*floatFunc)(), int nextra, struct MRCheader  *header,
                   FILE *infp, FILE *outfp, int direction, char *cmap);


int main( int argc, char *argv[])
{
  void  (*convertFunc)();
  void  (*floatFunc)();
  struct MRCheader hdata;
  struct MRCheader hconv;
  struct MRCheader *hout;
  int error;
  int   factor;
  int   size;
  int   tonative;
  int   swapieee;
  FILE  *fin, *fout;
  int   i;
  int   inplace;
  unsigned char bdata;
  int filesize, datasize;
  char *progname = imodProgName(argv[0]);

#ifdef SWAP_IEEE_FLOATS
  swapieee = TRUE;
#else
  swapieee = FALSE;
#endif

#ifndef B3D_LITTLE_ENDIAN
  if (argc < 2){
    fprintf(stderr, "Usage: %s [-vms|-ieee] filename [optional output filename]\n",
            progname);
#ifdef SWAP_IEEE_FLOATS
    fprintf(stderr, "options: choose only one, default is -ieee.\n");
#else
    fprintf(stderr, "options: choose only one, default is -vms.\n");
#endif
    fprintf(stderr, "\t-vms   convert between VMS and IEEE format\n");
    fprintf(stderr, "\t-ieee  swap bytes in IEEE format only\n");
    exit(1);
  }

  i = 1;
  if (argv[1][0] == '-'){
    if (argv[1][1] == 'v'){
      swapieee = FALSE;
      i = 2;
    }else{
      if (argv[1][1] == 'i'){
        swapieee = TRUE;
        i = 2;
      }
    }
  }

  if (i >= argc){
    fprintf(stderr, "%s: Usage [-vms|-ieee] [filename] [optional output filename]\n",
            progname);
    exit(1);
  }
#else
  if (argc < 2 || argc > 3 || argv[1][0] == '-') {
    fprintf(stderr, "%s: Usage [filename] [opt filename]\n",
            progname);
    fprintf(stderr, "There is no option to convert VMS data on this machine.\n");
    exit(1);
  }
  i = 1;
#endif
     
  fin = fopen(argv[i], "rb");
  if (fin == NULL){
    fprintf(stderr, "%s: Couldn't open %s.\n", progname, argv[i]);
    exit(10);
  }
  i++;

  if (i >= argc){
    inplace = TRUE;
    fout = fopen(argv[i-1], "rb+");
  }
  else{
    fout = fopen(argv[i], "wb");
    inplace = FALSE;
  }

  if (fout == NULL){
    fprintf(stderr, "%s: Couldn't open %s.\n", progname, argv[i]);
    exit(10);
  }

  if (swapieee)
    floatFunc = convertLongs;
  else
    floatFunc = convertFloats;

  if (fread(&hdata, 56, 4, fin) == 0){
    fprintf(stderr, "%s: Error Reading header.\n", progname);
    exit(3);
  }

  hdata.swapped = 0;

  /* Test for byte-swapped data with image size and the map numbers */
  if (mrc_test_size(&hdata)) {

    /* Mark data as swapped and do the swaps of critical data */
    hdata.swapped = 1;
    mrc_swap_longs(&hdata.nx, 10);
    mrc_swap_longs(&hdata.mapc, 3);
    mrc_swap_longs(&hdata.next, 1);
    mrc_swap_shorts(&hdata.nint, 4);
    mrc_swap_longs(&hdata.nlabl, 1);

  }
  if (mrc_test_size(&hdata) ||
      hdata.mode > 16 || hdata.mode < 0) {
    fprintf(stderr, "%s: This is not an MRC file, even after "
            "swapping bytes in header.\n", progname);
      
    exit(3);
  }

  hdata.headerSize = 1024;
  hdata.headerSize += hdata.next;
  hdata.sectionSkip = 0;
  tonative = hdata.swapped;
  datasize = hdata.nx * hdata.ny * hdata.nz;
  fseek(fin, 0, SEEK_END);
  filesize = ftell(fin);
  switch(hdata.mode){
  case MRC_MODE_BYTE:
    convertFunc = convertBytes;
    size = 1;
    factor = 1;
    break;
      
  case MRC_MODE_SHORT:
    datasize *= sizeof(b3dInt16);
    convertFunc = convertShorts;
    size = sizeof(b3dInt16);
    factor = 1;
    break;
      
  case MRC_MODE_FLOAT:
    datasize *= sizeof(b3dFloat);
    convertFunc = floatFunc;
    size = sizeof(b3dFloat);
    factor = 1;
    break;
      
  case MRC_MODE_COMPLEX_SHORT:
    datasize *= 2 * sizeof(b3dInt16);
    convertFunc = convertShorts;
    size = sizeof(b3dInt16);
    factor = 2;
    break;
      
  case MRC_MODE_COMPLEX_FLOAT:
    datasize *= 2 * sizeof(b3dFloat);
    convertFunc = floatFunc;
    size = sizeof(b3dFloat);
    factor = 2;
    break;
      
  case MRC_MODE_RGB:
    datasize *= 3;
    convertFunc = convertBytes;
    size = 1;
    factor = 3;
    break;

  default:
    fprintf(stderr, "%s: data type %d unsupported.\n",
            progname, hdata.mode);
    exit(3);
    break;
  }

  datasize += hdata.headerSize;
  if (filesize < datasize || filesize > datasize + 511){
    fprintf(stderr, "mrcx: Warning input file is %d bytes, "
            "expected size is %d.\n", filesize, datasize);
  }
  rewind(fin);

#ifdef B3D_LITTLE_ENDIAN
  if (tonative)
    printf("Converting big-endian IEEE to little-endian IEEE.\n");
  else
    printf("Converting little-endian IEEE to big-endian IEEE.\n");
#else
  if (swapieee) {
    if (tonative)
      printf("Converting little-endian IEEE to big-endian IEEE.\n");
    else
      printf("Converting big-endian IEEE to little-endian IEEE.\n");
  } else {
    if (tonative)
      printf("Converting little-endian VMS to big-endian IEEE.\n");
    else
      printf("Converting big-endian IEEE to little-endian VMS.\n");
  }
#endif


  convertHeader(floatFunc, hdata.next, &hconv, fin, fout, tonative, 
                &hdata.cmap[0]);

  fwrite(&hconv, 56, 4, fout);
     
  for (i = 0; i < MRC_NLABELS; i++)
    fwrite(hconv.labels[i], MRC_LABEL_SIZE, 1,fout);

  if (hdata.next)
    fwrite(hconv.symops, hdata.next, 1, fout);


  if (!inplace || convertFunc != convertBytes)
    convertBody(convertFunc, &hdata, size, factor, tonative, fin, fout);
     
  fclose(fin);
  if (!inplace)
    fclose(fout);
  exit(0);
  return(0);
}



/*****************************************************************************/



void convertBody(void (*convertFunc)(),
                 struct MRCheader *header,
                 int size,
                 int factor,
                 int direction,
                 FILE *infp,
                 FILE *outfp)

     /*sets up the image data conversion and writes the converted data to the 
       output file*/

{
  unsigned int lcv;       
  b3dUByte *scanLine;  /*converted scanline*/
  int   scans = 0;
  b3dUByte  pix;
  int i;

  if (size == 1){
    size = header->nx * header->ny * header->nz;
    for(i = 0; i < size; i++){
      if (! fread(&(pix), 1,1,infp)){
        perror("Error copying data ");
        return;
      }
      fwrite(&(pix), 1, 1, outfp);
    }
    return;
  }

  scanLine = (b3dUByte *)malloc(header->nx * size * factor);
  
  scans = header->ny * header->nz;

  for (lcv = 0; lcv < scans; lcv++)
    {
      (*convertFunc)(infp, scanLine, header->nx * factor, direction);
      fwrite(scanLine, size, header->nx * factor, outfp);
    }

  free(scanLine);

} /*convertBody*/



/*****************************************************************************/
void convertBytes(FILE  *fp,        /*file to read from*/
                  b3dUByte *buffer, /*buffer to write the converted values to*/
                  int   noBytes,    /*number of bytes to convert*/
                  int   direction)  /*conversion direction*/

     /*converts byte data.*/

{
  /* Read in data */

  if (fread(buffer, 1, noBytes, fp) != noBytes)
    {
      fprintf(stderr, "Read Failed While Reading Byte Quantities\n");
    }

} /*convertBytes*/



/*****************************************************************************/
void convertFloats(FILE  *fp,       /*file to read from*/
                   b3dUByte *buffer,/*buffer to write the converted values to*/
                   int   noFloats,  /*number of floats to convert*/
                   int   direction) /*conversion direction*/

     /*
       Converts a floating point number in vax floating point representation to IEEE
       representation and vica versa. A vax floating point no is bias 128 with
       exponents ranging from -127 to 127 the number is stored little endian. An IEEE
       float is point no is bias 127 with exponents ranging form -126 to 128, the
       number is typically stored big endian.
     */

{
  b3dUByte exp;
  b3dUInt32  lcv;
  b3dUInt32  noBytes;
  b3dUByte temp;   /*temporary place holder*/

  noBytes = noFloats * sizeof(b3dFloat);

  if (fread(buffer, 1, noBytes, fp) != noBytes)
    {
      fprintf(stderr, "Read Failed While Reading Float Quantities\n");
    }

  /*When converting from vax floating to IEEE floating, it must be noted 
   *that vax float exponents range from -127 to 127 whilst IEEE exponents
   *range from -126 to 128. We must therefore clamp -127 vax floats to 
   *-126 IEEE floats. This is done by not subtracting when the vax
   *exponent is -127 (1 bias 128). Note that because the vax mantissa is <= 0.5
   *and the IEEE mantissa is <= 1 we subtract 2 from the exponent to make up
   *for this. When converting from IEEE floats to vax floats, we must clamp 128
   *IEEE floats to 127 vax floats. 
   *In either case corrent byte swapping must be done for
   *converting from big endian to little endian and vice versa*/ 

  for (lcv = 0; lcv < noBytes; lcv += sizeof(b3dFloat))
    {
      if (direction) /* FROM_VAX */
        {
          if ((exp = (buffer[lcv+1] << 1) | (buffer[lcv] >> 7 & 0x01)) > 3 && 
              exp != 0)
            buffer[lcv+1] -= 1;
          else if (exp <= 3 && exp != 0)  /*must zero out the mantissa*/
            {
              /*we want manitssa 0 & exponent 1*/

              buffer[lcv] = 0x80;
              buffer[lcv+1] &= 0x80;
              buffer[lcv+2] = buffer[lcv+3] = 0;
            }
       
          temp = buffer[lcv];
          buffer[lcv] = buffer[lcv+1];
          buffer[lcv+1] = temp;
        }
      else  /*TO_VAX*/
        {
          if ((exp = (buffer[lcv] << 1) | (buffer[lcv+1] >> 7 & 0x01)) < 253 &&
              exp != 0)
            buffer[lcv] += 1;
          else if (exp >= 253) /*must also max out the exp & mantissa*/
            {
              /*we want manitssa all 1 & exponent 255*/

              buffer[lcv] |= 0x7F;
              buffer[lcv+1] = 0xFF; 
              buffer[lcv+2] = buffer[lcv+3] = 0xFF;
            }

          temp = buffer[lcv];
          buffer[lcv] = buffer[lcv+1];
          buffer[lcv+1] = temp;
        }
      
      if (((direction  && exp > 3) || 
           (!direction  && exp < 253)) && exp != 0)
        {
          temp = buffer[lcv+2];
          buffer[lcv+2] = buffer[lcv+3];
          buffer[lcv+3] = temp;
        }
    }

} /*convertFloats*/

/******************************************************************************/
void convertHeader(void (*floatFunc)(),
                   int nextra,
                   struct MRCheader  *header,  /*MRC header structure*/
                   FILE         *infp,        /*input file pointer*/
                   FILE         *outfp,       /*input file pointer*/
                   int direction,
                   char *cmap)

{
  int lcv;
  b3dUByte extra;

  convertLongs(infp, (b3dUByte *)&header->nx, 1, direction);
  convertLongs(infp, (b3dUByte *)&header->ny, 1, direction);
  convertLongs(infp, (b3dUByte*)&header->nz, 1, direction);
     
  convertLongs(infp, (b3dUByte*)&header->mode, 1, direction);

  convertLongs(infp, (b3dUByte *)&header->nxstart, 1, direction);
  convertLongs(infp, (b3dUByte *)&header->nystart, 1, direction);
  convertLongs(infp, (b3dUByte *)&header->nzstart, 1, direction);
  convertLongs(infp, (b3dUByte *)&header->mx, 1, direction);
  convertLongs(infp, (b3dUByte *)&header->my, 1, direction);
  convertLongs(infp, (b3dUByte *)&header->mz, 1, direction);
  floatFunc(infp, (b3dUByte *)&header->xlen, 1, direction);
  floatFunc(infp, (b3dUByte *)&header->ylen, 1, direction);
  floatFunc(infp, (b3dUByte *)&header->zlen, 1, direction);
  floatFunc(infp, (b3dUByte *)&header->alpha, 1, direction);
  floatFunc(infp, (b3dUByte *)&header->beta, 1, direction);
  floatFunc(infp, (b3dUByte *)&header->gamma, 1, direction);
  convertLongs(infp, (b3dUByte *)&header->mapc, 1, direction);
  convertLongs(infp, (b3dUByte *)&header->mapr, 1, direction);
  convertLongs(infp, (b3dUByte *)&header->maps, 1, direction);
  floatFunc(infp, (b3dUByte *)&header->amin, 1, direction);
  floatFunc(infp, (b3dUByte *)&header->amax, 1, direction);
  floatFunc(infp, (b3dUByte *)&header->amean, 1, direction);
  convertLongs(infp, (b3dUByte *)&header->ispg, 1, direction);
  /*     convertLongs(infp, (b3dUByte *)header->extra, 16, direction); */
  convertLongs(infp, (b3dUByte *)&header->next, 1, direction);
  convertShorts(infp, (b3dUByte *)&header->creatid, 1, direction);
  convertBytes(infp, (b3dUByte *)&header->blank, 30, direction);
  convertShorts(infp, (b3dUByte *)&header->nint, 4, direction);
  floatFunc(infp, (b3dUByte *)&header->min2, 6, direction);

  convertShorts(infp, (b3dUByte *)&header->idtype, 1, direction);
  convertShorts(infp, (b3dUByte *)&header->lens, 1, direction);

  convertShorts(infp, (b3dUByte *)&header->nd1, 1, direction);
  convertShorts(infp, (b3dUByte *)&header->nd2, 1, direction);
  convertShorts(infp, (b3dUByte *)&header->vd1, 1, direction);
  convertShorts(infp, (b3dUByte *)&header->vd2, 1, direction);

  floatFunc(infp, (b3dUByte *)header->tiltangles, 6, direction);

  if (cmap[0] != 'M' || cmap[1] != 'A' || cmap[2] != 'P') {
    /* Old-style header, swap wavelength then origin data */
    convertShorts(infp, (b3dUByte *)&header->xorg, 6, direction);
    floatFunc(infp, (b3dUByte *)&header->cmap[0], 3, direction);
    printf("Preserving old-style MRC header.\n");
  } else {
    /* new style, swap floats and retain cmap; flip stamp */
    floatFunc(infp, (b3dUByte *)&header->xorg, 1, direction);
    floatFunc(infp, (b3dUByte *)&header->yorg, 1, direction);
    floatFunc(infp, (b3dUByte *)&header->zorg, 1, direction);
    convertBytes(infp, (b3dUByte *)&header->cmap[0], 8, direction);
    floatFunc(infp, (b3dUByte *)&header->rms, 1, direction);
    header->stamp[0] = (header->stamp[0] == 17) ? 68 : 17;
  }
  convertLongs(infp, (b3dUByte *)&header->nlabl, 1, direction);

  for (lcv = 0; lcv < MRC_NLABELS; lcv++){
    convertBytes(infp, (b3dUByte *)header->labels[lcv], MRC_LABEL_SIZE, 
                 direction);  
    header->labels[lcv][MRC_LABEL_SIZE] = '\0';
  }

  /* THIS WILL NOT WORK WITH DATA CONSISTING OF INTEGERS AND REALS */
  if (nextra > 0){
    header->symops = (b3dUByte *)malloc(nextra);
    convertShorts(infp, (b3dUByte *)header->symops, 
                  nextra/2, direction);  
    if (nextra % 2)
      convertBytes(infp, (b3dUByte *)header->symops + nextra - 1, 1, 
                   direction);
  }
     
  header->fp = outfp;
     
} /*convertheader*/


/******************************************************************************/
void convertLongs(FILE  *fp,        /*file to read from*/
                  b3dUByte *buffer, /*buffer to write the converted values to*/
                  int   noLongs,    /*number of longwords to convert*/
                  int   direction)  /*conversion direction*/

     /*converts a vax longword to a "unix" long word*/

{
  b3dUInt32  lcv;
  b3dUInt32  noBytes;  /*no bytes to read*/
  unsigned int lw;        /*long word place holder*/

  /* Read in data */

  noBytes = noLongs * sizeof(b3dInt32);

  if (fread(buffer, 1, noBytes, fp) != noBytes)
    {
      fprintf(stderr, "mrcx: Read Failed While Reading Long Quantities\n");
    }

  /* DNM 3/16/01: discovered that the code here did nothing on a little-endian
     machine!  Switch to call library routine */
  mrc_swap_longs((int *)buffer, noLongs);

} /*convertLong*/

/******************************************************************************/
void convertShorts(FILE  *fp,       /*file to read from*/
                   b3dUByte *buffer,/*buffer to write the converted values to*/
                   int   noShorts,  /*number of shortwords to convert*/
                   int   direction) /*conversion direction*/

     /*converts a vax shortword to a "unix" shortword*/

{
  b3dUInt32   lcv;
  b3dUInt32   noBytes;  /*no bytes to read from the input file*/
  b3dUInt16   s;        /*short word place holder*/

  /* Read in data */

  noBytes = noShorts * sizeof(b3dInt16);

  if (fread(buffer, 1, noBytes, fp) != noBytes)
    {
      fprintf(stderr, "mrcx: Read Failed While Reading Short Quantities\n");
    }

  /* DNM 3/16/01: discovered that the code here did nothing on a little-endian
     machine!  Switch to call library routine */
  mrc_swap_shorts((b3dInt16 *)buffer, noShorts);

} /*convertShorts*/


